hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/types/HazelcastIntegerType.java

```diff
@@ -0,0 +1,421 @@
+ * was expanded and the type was promoted to the next wider integer type; in the
+ * last example the bit width was expanded, but there was no wider integer type
+ * to promote to and avoid a possible overflow.
+ * <p>
+ * The benefits of that bit width approach are: the smallest possible type is
+ * always selected to represent a result of a certain operation and it's always
+ * possible to tell from the selected type alone whether overflow checking is
+ * necessary while executing the operation.
+ */
+public final class HazelcastIntegerType extends BasicSqlType {
+
+    private static final Map<SqlTypeName, HazelcastIntegerType[]> TYPES = new HashMap<>();
+    private static final Map<SqlTypeName, HazelcastIntegerType[]> NULLABLE_TYPES = new HashMap<>();
+
+    static {
+        // Preallocate all possible types of all possible bit widths.
+
+        TYPES.put(TINYINT, new HazelcastIntegerType[Byte.SIZE + 1]);
+        TYPES.put(SMALLINT, new HazelcastIntegerType[Short.SIZE + 1]);
+        TYPES.put(INTEGER, new HazelcastIntegerType[Integer.SIZE + 1]);
+        TYPES.put(BIGINT, new HazelcastIntegerType[Long.SIZE + 1]);
+
+        for (Map.Entry<SqlTypeName, HazelcastIntegerType[]> entry : TYPES.entrySet()) {
+            SqlTypeName typeName = entry.getKey();
+            HazelcastIntegerType[] types = entry.getValue();
+
+            HazelcastIntegerType[] nullableTypes = new HazelcastIntegerType[types.length];
+            NULLABLE_TYPES.put(typeName, nullableTypes);
+
+            for (int i = 0; i < types.length; ++i) {
+                types[i] = new HazelcastIntegerType(typeName, false, i);
+                nullableTypes[i] = new HazelcastIntegerType(typeName, true, i);
+            }
+        }
+    }
+
+    private static final HazelcastIntegerType[] TYPES_BY_BIT_WIDTH = new HazelcastIntegerType[Long.SIZE + 1];
+    private static final HazelcastIntegerType[] NULLABLE_TYPES_BY_BIT_WIDTH = new HazelcastIntegerType[Long.SIZE + 1];
+
+    static {
+        // Build reverse mapping structures to map from a bit width to a
+        // preferred integer type.
+
+        for (int i = 0; i <= Long.SIZE; ++i) {
+            HazelcastIntegerType type;
+            HazelcastIntegerType nullableType;
+            if (i < Byte.SIZE) {
+                type = TYPES.get(TINYINT)[i];
+                nullableType = NULLABLE_TYPES.get(TINYINT)[i];
+            } else if (i < Short.SIZE) {
+                type = TYPES.get(SMALLINT)[i];
+                nullableType = NULLABLE_TYPES.get(SMALLINT)[i];
+            } else if (i < Integer.SIZE) {
+                type = TYPES.get(INTEGER)[i];
+                nullableType = NULLABLE_TYPES.get(INTEGER)[i];
+            } else {
+                type = TYPES.get(BIGINT)[i];
+                nullableType = NULLABLE_TYPES.get(BIGINT)[i];
+            }
+
+            TYPES_BY_BIT_WIDTH[i] = type;
+            NULLABLE_TYPES_BY_BIT_WIDTH[i] = nullableType;
+        }
+    }
+
+    private final int bitWidth;
```

Comment:
This might be a topic for a separate PR, but still - it seems that maximum possible value is a better candidate than bit width. Consider an example when we add value of the same type N times. Every PLUS operation will add an additional bit, thus, estimating the maximum value to `2^N * MAX_VALUE`, while in reality, the max value is only `N * MAX_VALUE`. This way we may avoid unnecessary expansions.

Reply:
Yes, and it also might help with overflow detection for floating point types.

Reply:
Can we have a ticket for this?

Reply:
done https://github.com/hazelcast/hazelcast/issues/17308